sdk/core/azure-core/src/datetime.cpp

```diff
@@ -863,6 +863,15 @@ std::string DateTime::ToString(DateFormat format, TimeFractionFormat fractionFor
         "Unrecognized date format (" + std::to_string(static_cast<int64_t>(format)) + ").");
   }
 
+  // Casting to unsigned int so that this check covers both positive and negative enum values that
+  // are invalid and out of range.
+  if (static_cast<uint64_t>(fractionFormat) > static_cast<uint64_t>(TimeFractionFormat::Truncate))
```

Comment:
Why is this check not in the Else statement at line 911? Why do we need to do it early?

Also, this is the advantage of using a `switch` statement on an enumeration - the compiler will check for illegal options in the switch statement because there are only a limited set of values that need to be checked. You can put the "throw" statement in the `default` label in the switch statement and let the compiler deal with ALL possible invalid values - no casting required.

Reply:
I agree having a switch/default is cleaner, but it mixes logic/main purpose of the method and input validation.

We don't _need_ to do it early, but having input argument validation up front makes the code easier to understand and avoids unnecessary deep nesting. And this way we avoid intertwining of the logic building a datetime before bailing mid-way because we saw an invalid input.

That said, I will evaluate the use of a switch statement, to see what the code transformation looks like, with the benefit of avoiding the cast.

Reply:
Looking at it some more, there's value in keeping the check upfront, before `ThrowIfUnsupportedYear();` is called.

I am not sure it is worth transforming the if-logic below, into a switch statement, just to avoid this cast. The current approach has a more deterministic "fail-fast" behavior based on input parameters. It, at least, fulfills the initial intent of the check/tests.

Thoughts?

Reply:
I could change this check, right at this point, to the following.
Does that resolve your concern?

```C++
  switch (fractionFormat)
  {
    case TimeFractionFormat::DropTrailingZeros:
    case TimeFractionFormat::AllDigits:
    case TimeFractionFormat::Truncate:
      break;
    default:
      throw std::invalid_argument(
          "Unrecognized time fraction format ("
          + std::to_string(static_cast<int64_t>(fractionFormat)) + ").");
  }
```